src/hotspot/cpu/riscv/riscv_v.ad

```diff
@@ -123,6 +123,114 @@ instruct storeV(vReg src, vmemA mem) %{
   ins_pipe(pipe_slow);
 %}
 
+// vector load mask
+
+instruct vloadmask(vRegMask dst, vReg src) %{
+  match(Set dst (VectorLoadMask src));
+  format %{ "vloadmask $dst, $src" %}
+  ins_encode %{
+    __ vsetvli(t0, x0, Assembler::e8);
+    __ vmsne_vx(as_VectorRegister($dst$$reg), as_VectorRegister($src$$reg), x0);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct vloadmask_masked(vRegMask dst, vReg src, vRegMask_V0 vmask) %{
+  match(Set dst (VectorLoadMask src vmask));
+  format %{ "vloadmask_masked $dst, $src, $vmask" %}
+  ins_encode %{
+    __ vsetvli(t0, x0, Assembler::e8);
+    __ vmsne_vx(as_VectorRegister($dst$$reg), as_VectorRegister($src$$reg), x0, Assembler::v0_t);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+// vector store mask
+
+instruct vstoremask(vReg dst, vRegMask_V0 src, immI size) %{
+  match(Set dst (VectorStoreMask src size));
+  format %{ "vstoremask $dst, $src" %}
+  ins_encode %{
+    __ vsetvli(t0, x0, Assembler::e8);
+    __ vmv_v_x(as_VectorRegister($dst$$reg), x0);
+    __ vmerge_vim(as_VectorRegister($dst$$reg), as_VectorRegister($dst$$reg), 1);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+// vector mask compare
+
+instruct vmaskcmp(vRegMask dst, vReg src1, vReg src2, immI cond) %{
+  predicate(Matcher::vector_element_basic_type(n) == T_BYTE ||
+            Matcher::vector_element_basic_type(n) == T_SHORT ||
+            Matcher::vector_element_basic_type(n) == T_INT ||
+            Matcher::vector_element_basic_type(n) == T_LONG);
+  match(Set dst (VectorMaskCmp (Binary src1 src2) cond));
+  format %{ "vmaskcmp $dst, $src1, $src2, $cond" %}
+  ins_encode %{
+    BasicType bt = Matcher::vector_element_basic_type(this);
+    uint length_in_bytes = Matcher::vector_length_in_bytes(this);
+    __ compare_integral_v(as_VectorRegister($dst$$reg), bt, length_in_bytes, as_VectorRegister($src1$$reg),
+                          as_VectorRegister($src2$$reg), (int)($cond$$constant));
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct vmaskcmp_masked(vRegMask dst, vReg src1, vReg src2, immI cond, vRegMask_V0 vmask, vReg tmp) %{
```

Comment:
I think we can introduce another new operand type (say 'vRegMaskNoV0') which excludes mask register 'v0' for 'dst' here and other places where 'v0' could not be used as the destination register for a masked vector instruction as required by the RVV spec. Then we could eliminate the use of 'tmp' register and 'vmv1r.v' instruction.

Also, I would like to further rename 'vRegMask_V0 vmask' into 'vRegMask_V0 v0'. The RVV spec says that the mask value used to control execution of a masked vector instruction is always supplied by vector register 'v0' for now.

Reply:
Fixed.